packages/api-graphql/src/internals/operations/indexQuery.ts

```diff
@@ -0,0 +1,182 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { AmplifyServer } from '@aws-amplify/core/internals/adapter-core';
+import {
+	initializeModel,
+	generateGraphQLDocument,
+	buildGraphQLVariables,
+	flattenItems,
+	authModeParams,
+	getCustomHeaders,
+} from '../APIClient';
+import {
+	AuthModeParams,
+	ClientWithModels,
+	ListArgs,
+	V6Client,
+	V6ClientSSRRequest,
+	GraphQLResult,
+	QueryArgs,
+} from '../../types';
+import {
+	ModelIntrospectionSchema,
+	SchemaModel,
+} from '@aws-amplify/core/internals/utils';
+import { model } from '@aws-amplify/data-schema/lib-esm/src';
+
+export type IndexMeta = {
+	queryField: string;
+	pk: string;
+	sk?: string[];
+};
+
+export function indexQueryFactory(
+	client: ClientWithModels,
+	modelIntrospection: ModelIntrospectionSchema,
+	model: SchemaModel,
+	indexMeta: IndexMeta,
+	context = false,
+) {
+	const indexQueryWithContext = async (
+		contextSpec: AmplifyServer.ContextSpec,
+		args: QueryArgs,
+		options?: ListArgs,
+	) => {
+		return _indexQuery(
+			client,
+			modelIntrospection,
+			model,
+			indexMeta,
+			{
+				...args,
+				...options,
+			},
+			contextSpec,
+		);
+	};
+
+	const indexQuery = async (args: QueryArgs, options?: ListArgs) => {
+		return _indexQuery(client, modelIntrospection, model, indexMeta, {
+			...args,
+			...options,
+		});
+	};
+
+	return context ? indexQueryWithContext : indexQuery;
+}
+
+async function processGraphQlResponse(
+	result: GraphQLResult<any>,
+	selectionSet: undefined | string[],
+	modelInitializer: (flattenedResult: any[]) => any[],
+) {
+	const { data, extensions } = result;
+
+	const [key] = Object.keys(data);
+
+	if (data[key].items) {
+		const flattenedResult = flattenItems(data)[key];
+
+		if (selectionSet) {
+			// don't initialize models if custom selection set specified; return resolved values
+			return {
+				data: flattenedResult,
+				nextToken: data[key].nextToken,
+				extensions,
+			};
+		} else {
+			const initialized = modelInitializer(flattenedResult);
+
+			return {
+				data: initialized,
+				nextToken: data[key].nextToken,
+				extensions,
+			};
+		}
```

Comment:
Would a ternary be lighter without losing clarity?
```suggestion
		return {
			data: selectionSet ? flattenedResult : modelInitializer(flattenedResult),
			nextToken: data[key].nextToken,
			extensions,
		};
```